File: host/comm.py
import logging
import os
import sys
import time

from abc import ABC, abstractmethod
from collections import namedtuple
from enum import IntEnum
from typing import Optional, Union

from ledgerwallet.client import LedgerClient
from ledgerwallet.transport import enumerate_devices
from ledgerwallet.utils import serialize


Apdu = namedtuple("Apdu", "status data")
logger = logging.getLogger("comm")
CLA = 0x12


class ApduCmd(IntEnum):
    REQUEST_PAGE = 0x6101
    REQUEST_HMAC = 0x6102
    REQUEST_PROOF = 0x6103
    COMMIT_PAGE = 0x6201
    COMMIT_HMAC = 0x6202
    SEND_BUFFER = 0x6301
    RECV_BUFFER = 0x6401
    EXIT = 0x6501
    FATAL = 0x6601
    REQUEST_MANIFEST = 0x6701
    REQUEST_APP_PAGE = 0x6801
    REQUEST_APP_HMAC = 0x6802


class CommClient(ABC):
    def exchange(self, ins: int, data=b"", p1=0, p2=0, cla=CLA) -> Apdu:
        apdu = bytes([cla, ins, p1, p2])
        apdu += serialize(data)

        response = self._exchange(apdu)

        status_word = int.from_bytes(response[-2:], "big")
        return Apdu(status_word, response[:-2])

    @abstractmethod
    def __enter__(self) -> "CommClient":
        pass

    @abstractmethod
    def __exit__(self, type, value, traceback):
        pass

    @abstractmethod
    def _exchange(self, data: bytes) -> bytes:
        pass


class CommClientUSB(CommClient):
    def __init__(self):
        self.client = None

    def __enter__(self) -> "CommClientUSB":
        devices = enumerate_devices()
        if len(devices) == 0:
            logger.error("No Ledger device has been found.")
            sys.exit(0)

        self.client = LedgerClient(devices[0], cla=CLA)
        self.number_exchanges = 0
        self.total_time = 0.0

        logger_ledgerwallet = logging.getLogger("ledgerwallet")
        logger_ledgerwallet.setLevel(logging.INFO)

        return self

    def __exit__(self, type, value, traceback):
        average = int((self.total_time * 1000) // self.number_exchanges)
        logger.info(f"{self.number_exchanges} exchanges in {self.total_time:.1f} s, ~{average} ms / exchange")

        self.client.close()
        self.client = None

    def _exchange(self, data: bytes) -> bytes:
        start = time.time()
        response = self.client.raw_exchange(data)
        end = time.time()

        self.number_exchanges += 1
        self.total_time += end - start

        return response


class CommClientBLE(CommClient):
    def __init__(self):
        from ble import disconnect_ble_client, exchange_ble, get_client_ble
        self.disconnect_ble_client = disconnect_ble_client
        self.exchange_ble = exchange_ble
        self.get_client_ble = get_client_ble
        self.client = None

    def __enter__(self) -> "CommClientBLE":
        self.client = self.get_client_ble()
        self.number_exchanges = 0
        self.total_time = 0.0
        return self

    def __exit__(self, type, value, traceback):
        average = int((self.total_time * 1000) // self.number_exchanges)
        logger.info(f"{self.number_exchanges} exchanges in {self.total_time:.1f} s, ~{average} ms / exchange")

        self.disconnect_ble_client(self.client)
        self.client = None

    def _exchange(self, data: bytes) -> bytes:
        start = time.time()
        response = self.exchange_ble(self.client, data)
        end = time.time()

        self.number_exchanges += 1
        self.total_time += end - start

        return response


client: Optional[Union[CommClientUSB, CommClientBLE]] = None


def get_client(transport="usb", use_speculos=False) -> Union[CommClientUSB, CommClientBLE]:
    global client

    if use_speculos and transport == "usb":
        os.environ["LEDGER_PROXY_ADDRESS"] = "127.0.0.1"
        os.environ["LEDGER_PROXY_PORT"] = "9999"

    if client is None:
        if transport.lower() == "usb":
            client = CommClientUSB()
        elif transport.lower() == "ble":
            client = CommClientBLE()
        else:
            assert False

    return client

File: host-rust/src/lib/comm.rs
use cpython::{ObjectProtocol, PyBytes, PyClone, PyObject, Python};
use std::convert::TryInto;

#[derive(Debug, Eq, PartialEq)]
pub enum Status {
    RequestPage = 0x6101,
    RequestHmac = 0x6102,
    RequestProof = 0x6103,
    CommitPage = 0x6201,
    CommitHmac = 0x6202,
    SendBuffer = 0x6301,
    RecvBuffer = 0x6401,
    Exit = 0x6501,
    Fatal = 0x6601,
    RequestManifest = 0x6701,
    RequestAppPage = 0x6801,
    RequestAppHmac = 0x6802,
    Success = 0x9000,
}

impl From<u16> for Status {
    fn from(n: u16) -> Self {
        match n {
            0x6101 => Status::RequestPage,
            0x6102 => Status::RequestHmac,
            0x6103 => Status::RequestProof,
            0x6201 => Status::CommitPage,
            0x6202 => Status::CommitHmac,
            0x6301 => Status::SendBuffer,
            0x6401 => Status::RecvBuffer,
            0x6501 => Status::Exit,
            0x6601 => Status::Fatal,
            0x6701 => Status::RequestManifest,
            0x6801 => Status::RequestAppPage,
            0x6802 => Status::RequestAppHmac,
            0x9000 => Status::Success,
            _ => panic!("invalid APDU status"),
        }
    }
}

pub struct Apdu<'a> {
    cla: u8,
    ins: u8,
    p1: u8,
    p2: u8,
    lc: u8,
    data: &'a [u8],
}

impl Apdu<'_> {
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = vec![0; 5 + self.data.len()];
        bytes[0] = self.cla;
        bytes[1] = self.ins;
        bytes[2] = self.p1;
        bytes[3] = self.p2;
        bytes[4] = self.lc;
        bytes[5..].copy_from_slice(self.data);
        bytes
    }
}

pub struct Comm {
    comm: PyObject,
}

impl Comm {
    pub fn new(py: Python, comm: &PyObject) -> Comm {
        Comm {
            comm: comm.clone_ref(py),
        }
    }

    pub fn exchange_apdu(&self, apdu: &[u8]) -> Vec<u8> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let data = PyBytes::new(py, apdu);
        let result: PyBytes = self
            .comm
            .call_method(py, "_exchange", (&data,), None)
            .unwrap()
            .extract(py)
            .unwrap();
        result.data(py).to_vec()
    }

    pub fn exchange(
        &self,
        ins: u8,
        data: &[u8],
        p1: Option<u8>,
        p2: Option<u8>,
        cla: Option<u8>,
    ) -> (Status, Vec<u8>) {
        let apdu = Apdu {
            cla: cla.unwrap_or(0x12),
            ins,
            p1: p1.unwrap_or(0x00),
            p2: p2.unwrap_or(0x00),
            lc: data.len().try_into().unwrap(),
            data,
        };
        let data = self.exchange_apdu(&apdu.to_bytes());
        assert!(data.len() >= 2);
        let status = u16::from_be_bytes(data[data.len() - 2..].try_into().unwrap());
        (Status::from(status), data[..data.len() - 2].to_vec())
    }
}
